# Ejercicio 1

## Resolución

In [1]:
import numpy as np
from interpolation import lagrange

### Los datos del problema

In [2]:
X = np.array([1960, 1970, 1990, 2000])
Y = np.array([3039585530, 3707475887, 5281653820, 6079603571])

estimaciones = {1980: 4452584592}

### Función auxiliar

Declaramos una función auxiliar para evitar repetición, siendo su argumento las posiciones de los datos considerados en la estimación (por ejemplo, para 1970 y 1990 serían las posiciones 1 y 2)

In [3]:
def calcular_error_estimacion(slice):
    P, _ = lagrange(X[slice], Y[slice])
    aprox = P(1980)
    real = estimaciones[1980]
    err_abs = abs(real - aprox)
    err_rel = err_abs/real
    print("Usando las estimaciones de:", X[slice])
    print(f"El valor aproximado es {aprox:.6f}, comparado con el valor estimado de {real:.6f} nos da un error de {err_abs:.6f} (aprox. {err_rel*100:.2f}%)")

### Letra a

Usando los datos de 1970 y 1990.

In [4]:
calcular_error_estimacion([1,2])

Usando las estimaciones de: [1970 1990]
El valor aproximado es 4494564853.499939, comparado con el valor estimado de 4452584592.000000 nos da un error de 41980261.499939 (aprox. 0.94%)


### Letra b

Usando los datos de 1960, 1970 y 1990.

In [5]:
calcular_error_estimacion([0,1,2])

Usando las estimaciones de: [1960 1970 1990]
El valor aproximado es 4454831983.656250, comparado con el valor estimado de 4452584592.000000 nos da un error de 2247391.656250 (aprox. 0.05%)


### Letra c

Usando todos los datos

In [6]:
calcular_error_estimacion([0,1,2,3])

Usando las estimaciones de: [1960 1970 1990 2000]
El valor aproximado es 4472888285.656250, comparado con el valor estimado de 4452584592.000000 nos da un error de 20303693.656250 (aprox. 0.46%)


## Conclusiones

- En este ejercicio, usamos la interpolación de Lagrange empleando diferentes valores para estimar la población en 1980.
- En la primera aproximación usamos 2 valores lo que resultó en un error relativo de 0.94% comparándolo con el valor estimado.
- Al aumentar a 3 los valores usados, el error relativo se redujo a 0.05%, lo cual representa una mejora significativa en la aproximación. Sin embargo, cuando aumentamos a 4 los valores usados el error relativo aumenta a 0.46% respecto a la iteración anterior.
- Podemos concluir que aumentar los valores usados en la interpolación no siempre aumenta su exactitud.